### DISCUSSION TITLE
Difference between `AsyncValue` factories and `AsyncValue` implementation classes

### DISCUSSION THREAD

**Discussion by om-ha** [rrousselGit/riverpod#1535] [Q&A] [answered]
What is difference between:

- A. `AsyncValue` factories:
```dart
AsyncValue<T>.data // e.g. `AsyncValue<T?>.data(null);`
AsyncValue<T>.loading // e.g. `const AsyncValue<MyType>.loading();`
AsyncValue<T>.error
```
- B. `AsyncValue` implementations:
```dart
AsyncData<T> // e.g. `AsyncData<T?>(null);`
AsyncError<T> // e.g. `const AsyncLoading<MyType>()`
AsyncLoading<T>
```

When to use one and another?
1. As an API developer, producing logic via generic classes, generic methods, and top-level providers.
2. As an App developer, consuming 1. in app logic (providers etc..) or app UI (widgets).
3. Which is the canonical item to use? It feels like using the implementation classes are the canonical version here (`AsyncData`, `AsyncError`, `AsyncLoading`) .

**Comment by om-ha** [comment]
It feels more consistent to use `AsyncValue` factories, to not deviate from the name `AsyncValue`.

**Comment by rrousselGit** [comment, marked as answer]
They are strictly the same thing.

**Reply by om-ha** [reply]
So there is not preference for one over the other.

Marking this as answered then.
